activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -844,11 +844,27 @@ def configure_connection
 
         defaults = [':default', :default].to_set
 
-        # Make MySQL reject illegal values rather than truncating or blanking them, see
-        # http://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_strict_all_tables
+        # The default SQL mode in MySQL 5.7 includes these modes:
+        #   ONLY_FULL_GROUP_BY, STRICT_TRANS_TABLES, NO_ZERO_IN_DATE, NO_ZERO_DATE,
+        #   ERROR_FOR_DIVISION_BY_ZERO, NO_AUTO_CREATE_USER, and NO_ENGINE_SUBSTITUTION.
+        # See https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sql-mode-setting
         # If the user has provided another value for sql_mode, don't replace it.
         unless variables.has_key?('sql_mode') || defaults.include?(@config[:strict])
-          variables['sql_mode'] = strict_mode? ? 'STRICT_ALL_TABLES' : ''
+          sql_mode = %w(
+            ONLY_FULL_GROUP_BY STRICT_TRANS_TABLES NO_ZERO_IN_DATE NO_ZERO_DATE
+            ERROR_FOR_DIVISION_BY_ZERO NO_AUTO_CREATE_USER NO_ENGINE_SUBSTITUTION
+          )
+          if strict_mode?
+            sql_mode += ['STRICT_ALL_TABLES']
+          else
+            sql_mode -= ['STRICT_TRANS_TABLES']
+          end
+          # Before 5.7.5, MySQL does not detect functional dependency and ONLY_FULL_GROUP_BY is not enabled by default.
+          # See https://dev.mysql.com/doc/refman/5.7/en/group-by-handling.html
+          if version < '5.7.5'
+            sql_mode -= ['ONLY_FULL_GROUP_BY']
+          end
```

Comment:
In any case, we'd get out of the business of deducing the correct sql mode for different versions of MySQL and MariaDB and others.